### PR TITLE
scheduler: add debug log

### PR DESCRIPTION
**Pull request by Connor1996**
this pr:
- add some debug log in balance-leader and balance-region scheduler
- add a log api for pd-ctl to set log level

### CODE REVIEW

File: server/schedulers/balance_leader.go [hunk 1]
```diff
@@ -74,6 +75,8 @@ func (l *balanceLeaderScheduler) Schedule(cluster schedule.Cluster, opInfluence
 
 	source := cluster.GetStore(region.Leader.GetStoreId())
 	target := cluster.GetStore(newLeader.GetStoreId())
+	log.Debugf("source store is %v", source)
```
Comment (siddontang) [inline server/schedulers/balance_leader.go]: please combine two logs into one, add scheduler name .

File: server/schedulers/utils.go [hunk 2]
```diff
@@ -39,6 +39,8 @@ func scheduleTransferLeader(cluster schedule.Cluster, schedulerName string, s sc
 
 	mostLeaderStore := s.SelectSource(cluster, stores, filters...)
 	leastLeaderStore := s.SelectTarget(cluster, stores, filters...)
+	log.Debugf("mostLeaderStore is %v", mostLeaderStore)
```
Comment (siddontang) [inline server/schedulers/utils.go]: ditto

File: server/schedulers/utils.go [hunk 3]
```diff
@@ -63,7 +67,7 @@ func scheduleTransferLeader(cluster schedule.Cluster, schedulerName string, s sc
 			region, peer = scheduleRemoveLeader(cluster, schedulerName, mostLeaderStore.GetId(), s)
 		}
 	}
-
+	log.Debugf("transfer leader select %v to be new leader for region %v", peer, region)
```
Comment (siddontang) [inline server/schedulers/utils.go]: `[region xxx] do ...`

File: server/schedulers/balance_leader.go (hunk 1 above)
Comment (siddontang) [reply server/schedulers/balance_leader.go]: add `[region xxx] xxx`

Comment (Connor1996) [reply server/schedulers/balance_leader.go]: ok, but I think it's not necessary to add scheduler name, cause log will print file name

File: server/api/debug.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	err = json.Unmarshal(data, &debug)
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	level := "info"
```
Comment (disksing) [inline server/api/debug.go]: I think it's better to set log level by string. If original level is "warning" or "error", we will not able to recover after debug.

File: server/schedule/filters.go [hunk 5]
```diff
@@ -31,6 +32,7 @@ type Filter interface {
 func FilterSource(opt Options, store *core.StoreInfo, filters []Filter) bool {
 	for _, filter := range filters {
 		if filter.FilterSource(opt, store) {
+			log.Debugf("[filter %T] filtes store %v from source", filter, store)
```
Comment (disksing) [inline server/schedule/filters.go]: filtes -> filters.

File: server/api/debug.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	if debug {
+		level = "debug"
+	}
+	h.svr.SetLogLevel(level)
```
Comment (siddontang) [inline server/api/debug.go]: I worry the data race here. Changing log level is not thread safe?

**Comment by siddontang** [comment]
please attach `[region %d]` at the beginning of the log if there is.

File: server/api/debug.go (hunk 6 above)
Comment (Connor1996) [reply server/api/debug.go]: server reads cfg.Log only when initing logger, and setloglevel is not an operation with high frequency. Even if it causes write-write-conflict, it seems that it does not matter here.

File: server/schedulers/balance_leader.go [hunk 7]
```diff
@@ -74,6 +75,7 @@ func (l *balanceLeaderScheduler) Schedule(cluster schedule.Cluster, opInfluence
 
 	source := cluster.GetStore(region.Leader.GetStoreId())
 	target := cluster.GetStore(newLeader.GetStoreId())
+	log.Debugf("source store is %v, target store is %v", source, target)
```
Comment (disksing) [inline server/schedulers/balance_leader.go]: log store id should be enough.

File: server/schedulers/balance_region.go [hunk 8]
```diff
@@ -117,6 +118,7 @@ func (s *balanceRegionScheduler) transferPeer(cluster schedule.Cluster, region *
 
 	target := cluster.GetStore(newPeer.GetStoreId())
 	avgScore := cluster.GetStoresAverageScore(core.RegionKind)
+	log.Debugf("source store is %v, target store is %v", source, target)
```
Comment (disksing) [inline server/schedulers/balance_region.go]: log store id should be enough.

File: server/schedule/filters.go [hunk 9]
```diff
@@ -31,6 +32,7 @@ type Filter interface {
 func FilterSource(opt Options, store *core.StoreInfo, filters []Filter) bool {
 	for _, filter := range filters {
 		if filter.FilterSource(opt, store) {
+			log.Debugf("[filter %T] filters store %v from source", filter, store)
```
Comment (siddontang) [inline server/schedule/filters.go]: %T will output all the fields of the filter, is it ok here?

File: server/schedulers/balance_leader.go [hunk 10]
```diff
@@ -74,6 +75,7 @@ func (l *balanceLeaderScheduler) Schedule(cluster schedule.Cluster, opInfluence
 
 	source := cluster.GetStore(region.Leader.GetStoreId())
 	target := cluster.GetStore(newLeader.GetStoreId())
+	log.Debugf("source store id is %v, target store id is %v", source.GetId(), target.GetId())
```
Comment (siddontang) [inline server/schedulers/balance_leader.go]: add region

File: server/schedulers/balance_region.go [hunk 11]
```diff
@@ -117,6 +118,7 @@ func (s *balanceRegionScheduler) transferPeer(cluster schedule.Cluster, region *
 
 	target := cluster.GetStore(newPeer.GetStoreId())
 	avgScore := cluster.GetStoresAverageScore(core.RegionKind)
+	log.Debugf("source store id is %v, target store id is %v", source.GetId(), target.GetId())
```
Comment (siddontang) [inline server/schedulers/balance_region.go]: add region

File: server/schedulers/utils.go [hunk 12]
```diff
@@ -39,6 +39,7 @@ func scheduleTransferLeader(cluster schedule.Cluster, schedulerName string, s sc
 
 	mostLeaderStore := s.SelectSource(cluster, stores, filters...)
 	leastLeaderStore := s.SelectTarget(cluster, stores, filters...)
+	log.Debugf("[%v] mostLeaderStore is %v, leastLeaderStore is %v", schedulerName, mostLeaderStore, leastLeaderStore)
```
Comment (siddontang) [inline server/schedulers/utils.go]: `[%s]`

File: server/schedule/filters.go (hunk 9 above)
Comment (Connor1996) [reply server/schedule/filters.go]: no, it outputs like "*schedule.excludedFilter"

File: server/schedulers/utils.go [hunk 13]
```diff
@@ -63,7 +65,11 @@ func scheduleTransferLeader(cluster schedule.Cluster, schedulerName string, s sc
 			region, peer = scheduleRemoveLeader(cluster, schedulerName, mostLeaderStore.GetId(), s)
 		}
 	}
-
+	if region == nil {
+		log.Debugf("[%v] select no region", schedulerName)
+	} else {
+		log.Debugf("[region %v][%v] select %v to be new leader", schedulerName, region.GetId(), peer)
```
Comment (nolouch) [inline server/schedulers/utils.go]: I guess the parameter does not correspond.

File: server/schedulers/utils.go [hunk 14]
```diff
@@ -183,6 +190,7 @@ func shouldBalance(source, target *core.StoreInfo, avgScore float64, kind core.R
 	sourceSizeDiff := (sourceScore - avgScore) * source.ResourceWeight(kind)
 	targetSizeDiff := (avgScore - targetScore) * target.ResourceWeight(kind)
 
+	log.Debugf("size diff is %v and region size is %v", math.Min(sourceSizeDiff, targetSizeDiff), float64(region.ApproximateSize)*tolerantRatio)
```
Comment (nolouch) [inline server/schedulers/utils.go]: region size -> tolerable size？better to indicate region id.

File: server/schedulers/utils.go (hunk 13 above)
Comment (Connor1996) [reply server/schedulers/utils.go]: oh, I mistake the order

**Comment by siddontang** [comment]
LGTM

**Pull request merged**